### PR TITLE
Rationalise the Python-specific options

### PR DESCRIPTION
**Pull request by ojwb**
This is a significant start on cleaning out the cruft ready for 4.0.0 as agreed in #701.  I haven't yet started to rationalise the command line options, but with this done we can make `-modern` the default and drop support for `-classic`, and then hopefully pick away at some of the others.

It removes some of the function casts which GCC8 warns about (#1259), but is not yet a complete fix for that.

### CODE REVIEW

File: Source/Modules/python.cxx [hunk 1]
```diff
@@ -836,7 +834,7 @@ class PYTHON:public Language {
       Printf(default_import_code, tab8 tab4 "return importlib.import_module('%s')\n", module);
       Printf(default_import_code, tab4 "%s = swig_import_helper()\n", module);
       Printv(default_import_code, tab4, "del swig_import_helper\n", NULL);
-      Printv(default_import_code, "elif _swig_python_version_info >= (2, 6, 0):\n", NULL);
```
Comment (wsfulton) [inline Source/Modules/python.cxx]: I think we should raise a RuntimeError if < 2.6 saying unsupported version to give a helpful error message to users trying to use older versions.

Also the C code should `#error` if < 2.6.

**Comment by ojwb** [comment]
>
> I think we should raise a RuntimeError if < 2.6 saying unsupported version
> to give a helpful error message to users trying to use older versions.
>
OK.

> Also the C code should #error if < 2.6.
>
It already does.

>

File: Source/Modules/python.cxx (hunk 1 above)
Comment (wsfulton) [reply Source/Modules/python.cxx]: Ah, I just saw elsewhere you have already put in the check in the C code, so ignore that comment

**Comment by wsfulton** [comment]
Hey @ojwb, please don't get carried away modifying python.cxx, the doxygen code needs merging first, which I'm working on now. Also, I think focus should be on getting -O working before removing anything more because that is what our eventual aim is. Until that is fully working, we can't be confident that we can just remove any of the non -O options.

**Comment by ojwb** [comment]
This PR is blocked waiting for the doxygen branch to be merged, but meanwhile we're wasting a lot of travis resources on 6 CI builds for Python versions we're not intending to support, and so I've pushed a change to remove those builds (f7a255c38acff21cad01e300ba8d8169aea9c377) - this shouldn't cause problematic conflicts with the doxygen branch.

But that's why this PR now shows a conflict on `.travis.yml`.  I'll resolve that when I pick up work on this again.

**Comment by ojwb** [comment]
The doxygen branch is now merged, so I've picked out the changes to drop support for Python < 2.7 and Python 3 < 3.4 and pushed those to master, then rebased this branch on current master so it just has the changes towards unpicking the thicket of Python options which we want to hack back.

**Comment by mromberg** [comment]
> AIUI, the only area where -O doesn't currently work is with the nondynamic feature, which there seemed to be a firm consensus on removing However when I said I was actually going to remove it you told me to hold off on doing so.
> 
> I view the current incompatibility as being a problem with nondynamic rather than with -O, and I really don't see the point in investing effort in trying to fix a feature which seems to have no users and which nobody likes.

  Yes.   The nondynamic feature was one of the last things I tried to tackle when getting all the -O tests to work.   There is still an open pull request with my changes #718.   These changes did make the tests that failed pass at the time.   However, I don't think #718 should be applied at all.   It is mostly useful as a reference.  My opinion is that the nondynamic feature should be removed all together.   

  It makes the most sense to remove the nondynamic tests first.  This would as a side effect make -O also work.   Then the more detailed surgery of removing nondynamic and all the complicated meta class stuff can occur.

  It is very hard to imagine any working swig generated code that would break if swig were to remove the nondynamic feature.  To be safe it could be made a no op and or issue a warning.  

  What nondynamic does is cause python to throw an exception if any python code attempts to add new attributes to a swig generated class.   So, almost all user code that may have enabled nondynamic would not be doing this (adding new attributes).   The only code that would break is I suppose code that was expecting these exceptions and catching them.   And while I can't say for certain no one is doing this.   It would surprise me if anyone were.

  So, I think the tests for nondynamic should just be removed.   I also think the feature itself should be removed unless any super convincing argument can be made for it's existence.   The argument against it is that there is a lot of questionable code inserted into python wrappers to support it.  And this questionable code is being used even if the feature nondynamic is never turned on.

**Comment by wsfulton** [comment]
Now #1293 needs addressing before we can focus on the full rationalisation of the Python options. Please add any thoughts about type errors to #1293.

**Comment by wsfulton** [comment]
With #1293 fixed, I am now working on merging this patch into master. There are a number of conflicts, but I think I have it all working now. I'll do some more testing, then commit it tomorrow.

**Comment by wsfulton** [comment]
@ojwb, thanks for the patch, closing.

**Pull request closed without merging**